File: packages/py-moose-lib/moose_lib/dmv2/sql_resource.py
"""
Base SQL resource definitions for Moose Data Model v2 (dmv2).

This module provides the base class for SQL resources like Views and Materialized Views,
handling common functionality like setup/teardown SQL commands and dependency tracking.
"""
from typing import Any, Optional, Union, List
from pydantic import BaseModel

from .olap_table import OlapTable
from ._registry import _sql_resources

class SqlResource:
    """Base class for SQL resources like Views and Materialized Views.

    Handles the definition of setup (CREATE) and teardown (DROP) SQL commands
    and tracks data dependencies.

    Attributes:
        name (str): The name of the SQL resource (e.g., view name).
        setup (list[str]): SQL commands to create the resource.
        teardown (list[str]): SQL commands to drop the resource.
        pulls_data_from (list[SqlObject]): List of tables/views this resource reads from.
        pushes_data_to (list[SqlObject]): List of tables/views this resource writes to.
        kind: The kind of the SQL resource (e.g., "SqlResource").
    """
    setup: list[str]
    teardown: list[str]
    name: str
    kind: str = "SqlResource"
    pulls_data_from: list[Union[OlapTable, "SqlResource"]]
    pushes_data_to: list[Union[OlapTable, "SqlResource"]]

    def __init__(
            self,
            name: str,
            setup: list[str],
            teardown: list[str],
            pulls_data_from: Optional[list[Union[OlapTable, "SqlResource"]]] = None,
            pushes_data_to: Optional[list[Union[OlapTable, "SqlResource"]]] = None,
            metadata: dict = None
    ):
        self.name = name
        self.setup = setup
        self.teardown = teardown
        self.pulls_data_from = pulls_data_from or []
        self.pushes_data_to = pushes_data_to or []
        self.metadata = metadata
        _sql_resources[name] = self

File: packages/ts-moose-lib/src/dmv2/sdk/sqlResource.ts
import { getMooseInternal } from "../internal";
import { OlapTable } from "./olapTable";
import { Sql, toStaticQuery } from "../../sqlHelpers";

type SqlObject = OlapTable<any> | SqlResource;

/**
 * Represents a generic SQL resource that requires setup and teardown commands.
 * Base class for constructs like Views and Materialized Views. Tracks dependencies.
 */
export class SqlResource {
  /** @internal */
  public readonly kind = "SqlResource";

  /** Array of SQL statements to execute for setting up the resource. */
  setup: readonly string[];
  /** Array of SQL statements to execute for tearing down the resource. */
  teardown: readonly string[];
  /** The name of the SQL resource (e.g., view name, materialized view name). */
  name: string;

  /** List of OlapTables or Views that this resource reads data from. */
  pullsDataFrom: SqlObject[];
  /** List of OlapTables or Views that this resource writes data to. */
  pushesDataTo: SqlObject[];

  /**
   * Creates a new SqlResource instance.
   * @param name The name of the resource.
   * @param setup An array of SQL DDL statements to create the resource.
   * @param teardown An array of SQL DDL statements to drop the resource.
   * @param options Optional configuration for specifying data dependencies.
   * @param options.pullsDataFrom Tables/Views this resource reads from.
   * @param options.pushesDataTo Tables/Views this resource writes to.
   */
  constructor(
    name: string,
    setup: readonly (string | Sql)[],
    teardown: readonly (string | Sql)[],
    options?: {
      pullsDataFrom?: SqlObject[];
      pushesDataTo?: SqlObject[];
    },
  ) {
    const sqlResources = getMooseInternal().sqlResources;
    if (sqlResources.has(name)) {
      throw new Error(`SqlResource with name ${name} already exists`);
    }
    sqlResources.set(name, this);

    this.name = name;
    this.setup = setup.map((sql) =>
      typeof sql === "string" ? sql : toStaticQuery(sql),
    );
    this.teardown = teardown.map((sql) =>
      typeof sql === "string" ? sql : toStaticQuery(sql),
    );
    this.pullsDataFrom = options?.pullsDataFrom ?? [];
    this.pushesDataTo = options?.pushesDataTo ?? [];
  }
}
